### PR TITLE
GLPI Integration by SecInfra

### PR DESCRIPTION
**Pull request by SecInfra**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [x] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by SecInfra** [comment]
about the docker image there is the PR : https://github.com/demisto/dockerfiles/pull/7471

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+                if 'Mirrored from Cortex XSOAR' not in note.get('content') and note.get('date_mod') > formated_date:
+                    comments_context = {'comments_and_work_notes': unescape(note.get('content'))}
+                    entries.append({
+                        'ContentsFormat': formats['html'],
+                        'Type': entryTypes['note'],
+                        'Contents': unescape(note.get('content')),
+                        'Note': True,
+                        'EntryContext': comments_context
+                    })
+        demisto.debug(f'Pull result is {new_incident_data}')
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/GLPI.py:816]: You meant to add the `entries` too?

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+        return GetRemoteDataResponse(new_incident_data, entries)
+    except Exception as e:
+        raise DemistoException('Error : ' + str(e))
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/GLPI.py:819]:
```suggestion
        raise DemistoException(f'Error in incoming mirror for incident id {ticket_id}. :Error message: {str(e)}')
```

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+
+    :rtype: ``str``
+    """
+    parsed_args = UpdateRemoteSystemArgs(args)
+    ticket_id = parsed_args.remote_incident_id
+    if parsed_args.delta:
+        demisto.debug(f'Got the following delta keys {str(list(parsed_args.delta.keys()))}')
+
+    demisto.debug(f'Sending incident with remote ID [{parsed_args.remote_incident_id}] to remote system\n')
+    new_incident_id: str = parsed_args.remote_incident_id
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/GLPI.py:847]: What is the difference between `ticket_id` and `new_incident_id`?

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 4]
```diff
@@ -0,0 +1,973 @@
...
+    updated_incident = {}
+
+    if not parsed_args.remote_incident_id or parsed_args.incident_changed:
+        if parsed_args.remote_incident_id:
+            old_incident = client.get_ticket(parsed_args.remote_incident_id)
+            for changed_key in parsed_args.delta.keys():
+                old_incident[changed_key] = parsed_args.delta[changed_key]  # type: ignore
+            parsed_args.data = old_incident
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/GLPI.py:855]: Notice that `parsed_args.data` holds everything that passed the mapper, and `parsed_args.delta` holds all the fields that were changed. `delta` can hold some server fields that have changed (for example, `closeReason` that appears when closing an incident in XSOAR and proving a reason for it). 

Do you want to take the values that are in `delta`, or the ones that are in `data`?
Maybe holding a list that has the names of the fields you want to take will be a good solution.
If `update_ticket` uses only the relevant fields then it is not a real problem.

File: Packs/GLPI/Integrations/GLPI/test/test.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+testing
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/test/test.txt:1]: Please move this to the `test_data` directory.

File: Packs/GLPI/Scripts/GLPIIncidentStatus/GLPIIncidentStatus.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+"""GLPIIncidentStatus Script for Cortex XSOAR (aka Demisto)"""
+
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+
+# from typing import Dict, Any
```
Comment (BEAdi) [inline Packs/GLPI/Scripts/GLPIIncidentStatus/GLPIIncidentStatus.py:7]:
```suggestion

```

**Review by BEAdi — CHANGES_REQUESTED** [review]
Hi,
Thank you for your large contribution!
The review is a bit long because I made it as comprehensive as possible, if there are any questions contact me.

Please mind the following:
1. Capitalize the names of the incident fields you introduced.
2. In the fetch mechanism, you did not take into account the `max_fetch` parameter. You should support it and limit the number of incidents fetched every time.
This can be done by sorting the incidents by time and returning only the first `max_fetch` ones (and updating the `last_fetch` accordingly of course). If there is a way that the items will return from `list_incidents` ordered by time, you can return the first `max_fetch` items that were returned from it.
3. In the `yml` file, add `defaultclassifier`, `defaultmapperin`, `defaultmapperout` with their ids.
4. I'm not sure how the `warnings.warn` you added in the code will behave, it will be better to change them to our log using `demisto.warning`.

File: Packs/GLPI/IncidentFields/incidentfield-GLPI_Category.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    "locked": false,
+    "name": "GLPI Category",
+    "neverSetAsRequired": false,
+    "openEnded": false,
+    "ownerOnly": false,
+    "required": false,
+    "sla": 0,
+    "system": false,
+    "threshold": 72,
+    "type": "shortText",
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/IncidentFields/incidentfield-GLPI_Category.json:24]: data from glpi, data which can be modified by users into glpi interface, no way to predict the values

File: Packs/GLPI/IncidentFields/incidentfield-GLPI_Type.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    "closeForm": false,
+    "content": true,
+    "description": "GLPI ticket type",
+    "editForm": true,
+    "group": 0,
+    "hidden": false,
+    "id": "incident_glpitype",
+    "isReadOnly": false,
+    "locked": false,
+    "name": "GLPI Type",
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/IncidentFields/incidentfield-GLPI_Type.json:17]: 2 incident types managed now : "GLPI Incident" and "GLPI Request"

File: Packs/GLPI/Integrations/GLPI/GLPI_description.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+Generate the **Application Token** :
+
+1 - On the Setup / General / API page, click on the "Add API client" button.
+2 - Ensure to set the Active field to yes and click the regenerate checkbox.
+3 - Once created, you can copy the Application token from the item page. 
+
+Generate the **User Token**
+
+1 -   Visit your profile preference page.
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI_description.md:22]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 10]
```diff
@@ -0,0 +1,973 @@
...
+                        'Note': True,
+                        'EntryContext': comments_context
+                    })
+        demisto.debug(f'Pull result is {new_incident_data}')
+        return GetRemoteDataResponse(new_incident_data, entries)
+    except Exception as e:
+        raise DemistoException('Error : ' + str(e))
+
+
+def update_remote_system_command(client: Client, args: Dict[str, Any], params: Dict[str, Any]) -> str:
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:822]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 11]
```diff
@@ -0,0 +1,973 @@
...
+                ticket_fields[arg] = input_arg
+    return ticket_fields
+
+
+def output_format(res, output_type=None, readable=None):
+    if res:
+        if isinstance(res, list):
+            keys = res[0].keys()
+        else:
+            keys = res.keys()
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:396]: when it's str (only one result), we directly use CommandResults

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 12]
```diff
@@ -0,0 +1,973 @@
...
+        key_list = []
+        for key in keys:
+            key_list.append(key)
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:399]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 13]
```diff
@@ -0,0 +1,973 @@
...
+    result = output_format(res, 'User', 'User with ID ' + str(args.get('id')) + ' successfully updated')
+    return result
+
+
+def delete_user_command(client, args):
+    username = args.get('name')
+    purge = args.get('purge')
+    user_id = get_user_id_helper(client, args)
+    client.delete_user(user_id, purge)
+    result = 'User ' + str(username) + ' successfully deleted'
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:535]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 14]
```diff
@@ -0,0 +1,973 @@
...
+def add_comment_command(client, args):
+    ticket_id = args.get('ticket_id')
+    text = args.get('comment')
+    res = client.add_comment(ticket_id, text)
+    if res:
+        if "id" in res[0]:
+            result = output_format(res, 'Comment', 'Comment successfully added to ticket ID : ' + str(ticket_id))
+            return result
+    else:
+        raise DemistoException('Error add comment')
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:564]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 15]
```diff
@@ -0,0 +1,973 @@
+# import demistomock as demisto  # noqa: F401
+# from CommonServerPython import *  # noqa: F401
+
+import demistomock as demisto
+from CommonServerPython import *  # noqa # pylint: disable=unused-wildcard-import
+from CommonServerUserPython import *  # noqa
+
+
+# /docker_image_create name=glpi-api dependencies=glpi-api,bcrypt
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:9]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 16]
```diff
@@ -0,0 +1,973 @@
...
+    Returns:
+        'ok' if test passed, anything else will fail the test.
+    """
+
+    try:
+        client = Client(params)
+        result = client.test()
+        if 'valid_id' in result:
+            return 'ok'
+        return 'Test Failed!'
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:299]: internal mechanism of GLPI : Add message for failed test

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 17]
```diff
@@ -0,0 +1,973 @@
...
+    requester = []
+    assigned = []
+    watcher = []
+    users = client.get_ticket_users(ticket_id)
+    for user in users:
+        if user['type'] == 1:
+            requester.append(user['users_id'])
+        elif user['type'] == 2:
+            assigned.append(user['users_id'])
+        elif user['type'] == 3:
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:339]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 18]
```diff
@@ -0,0 +1,973 @@
...
+
+
+def get_ticket_docs_helper(client, ticket_id):
+    docs = client.get_ticket_docs(ticket_id)
+    files = []
+    if docs:
+        for doc in docs:
+            doc_name = client.get_item('Document', doc['documents_id']).get('filename')
+            file = client.download_document(doc['documents_id'], filename=doc_name)
+            filepath, filename = os.path.split(file)
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:366]: Not the same, doc_name is like a "display name" for the filename in glpi

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 19]
```diff
@@ -0,0 +1,973 @@
...
+    last_fetch = last_run.get('last_fetch')
+
+    # Handle first time fetch
+    if last_fetch is None:
+        last_fetch = dateparser.parse(first_fetch_time)
+    else:
+        last_fetch = dateparser.parse(last_fetch)
+
+    latest_created_time = dateparser.parse(last_fetch.strftime('%Y-%m-%d %H:%M:%S'))
+    search_date = last_fetch.strftime('%Y-%m-%d %H:%M:%S')
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:703]: it's the date format of glpi

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 20]
```diff
@@ -0,0 +1,973 @@
...
+        requester, assigned, watcher = get_ticket_users_helper(client, ticket_id)
+        ticket['requester_users'] = requester
+        ticket['assigned_users'] = assigned
+        ticket['watcher_users'] = watcher
+        grequester, gassigned, gwatcher = get_ticket_groups_helper(client, ticket_id)
+        ticket['requester_groups'] = grequester
+        ticket['assigned_groups'] = gassigned
+        ticket['watcher_groups'] = gwatcher
+        htmldetails = unescape(ticket['content'])
+        ticket['content'] = htmldetails
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:719]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 21]
```diff
@@ -0,0 +1,973 @@
...
+            })
+        incident_created_time = dateparser.parse(ticket['date'])
+        ticket['mirror_direction'] = MIRROR_DIRECTION.get(demisto.params().get('mirror_direction'))
+        ticket['mirror_instance'] = demisto.integrationInstance()
+        ticket['mirror_tags'] = [
+            demisto.params().get('comment_tag'),
+            demisto.params().get('file_tag'),
+            demisto.params().get('work_notes_tag')
+        ]
+        demisto.debug('incident occured : ' + str(incident_created_time.strftime(DATE_FORMAT)))  # type: ignore[union-attr]
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:735]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 22]
```diff
@@ -0,0 +1,973 @@
...
+
+
+def get_mapping_fields_command() -> GetMappingFieldsResponse:
+    """
+    Returns the list of fields for an incident type.
+    Args:
+        client: XSOAR client to use
+    Returns: Dictionary with keys as field names
+    """
+    incident_type_scheme = SchemeTypeMapping(type_name="incident")
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:758]: I've changed it, but how to manage it now that we have 2 incident types ? an array ?

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 23]
```diff
@@ -0,0 +1,973 @@
...
+            lastUpdate: when was the last time we retrieved data
+    Returns:
+        List[Dict[str, Any]]: first entry is the incident (which can be completely empty) and the new entries.
+    """
+    parsed_args = GetRemoteDataArgs(args)
+    # ticket_id = args.get('id', '')
+    ticket_id = parsed_args.remote_incident_id
+    demisto.debug(f'Getting update for remote {ticket_id}')
+    last_update = args.get('lastUpdate')
+    demisto.debug('last_update: ' + str(last_update))
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:785]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 24]
```diff
@@ -0,0 +1,973 @@
...
+    formated_date = last_update.replace('T', ' ').split('.')[0]
+    try:
+        new_incident_data = client.get_ticket(ticket_id)
+        entries = []
+        demisto.debug('fetch files')
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:790]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 25]
```diff
@@ -0,0 +1,973 @@
...
+        ticket_docs = client.get_ticket_docs(ticket_id)
+        if ticket_docs:
+            for ticket_doc in ticket_docs:
+                if ticket_doc.get('date_mod') > formated_date:
+                    document = client.get_item('Document', ticket_doc.get('documents_id'))
+                    if '_mirrored_from_xsoar' not in document.get('filename'):
+                        file = client.download_document(ticket_doc.get('documents_id'), filename=document.get('filename'))
+                        demisto.debug('file fetched')
+                        filepath, filename = os.path.split(file)
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:799]: not the same values

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 2 above)
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:819]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 3 above)
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:847]: remove ticket_id

File: Packs/GLPI/Scripts/GLPIIncidentStatus/GLPIIncidentStatus.py (hunk 6 above)
Comment (SecInfra) [reply Packs/GLPI/Scripts/GLPIIncidentStatus/GLPIIncidentStatus.py:7]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 26]
```diff
@@ -0,0 +1,973 @@
...
+    glpi_pass = bpass.replace('$2b', '$2y')
+    additional_fields = split_fields(str(args.get('additional_fields', '')), ';')
+    profile_id = get_profile_id_helper(client, args)
+
+    user = {'name': username,
+            'realname': lastname,
+            '_useremails': [email],
+            'firstname': firstname,
+            'password': glpi_pass,
+            '_profiles_id': profile_id}
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:511]: I'm not sure to understand what you mean in that case.

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 1 above)
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:816]: Yes, to link it with the ticket

File: Packs/GLPI/Integrations/GLPI/test/test.txt (hunk 5 above)
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/test/test.txt]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py]: You're right, I make a list for it, thanks :)

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py]: done, thanks for the tip :)

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 27]
```diff
@@ -0,0 +1,1008 @@
...
+            'rawJSON': json.dumps(ticket)
+        }
+        incidents.append(incident)
+        count_incidents += 1
+
+        # Update last run and add incident if the incident is newer than last fetch
+        if incident_created_time > latest_created_time:  # type: ignore[operator]
+            latest_created_time = incident_created_time
+
+        if count_incidents >= max_results:
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/GLPI.py:764]: You can use `len(incidents)` instead

**Review by BEAdi — CHANGES_REQUESTED** [review]
Hi,
Thank you for the changes you did.
I added a few more changes requests.

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 28]
```diff
@@ -0,0 +1,1008 @@
...
+
+    incident_type_scheme = SchemeTypeMapping(type_name="GLPI Incident")
+    for field in GLPI_ARGS:
+        incident_type_scheme.add_field(field)
+    mapping_response.add_scheme_type(incident_type_scheme)
+
+    incident_type_scheme = SchemeTypeMapping(type_name="GLPI Request")
+    for field in GLPI_ARGS:
+        incident_type_scheme.add_field(field)
+    mapping_response.add_scheme_type(incident_type_scheme)
```
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:790]: My bad, a typo :)

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 27 above)
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py:764]: count_incidents removed

File: Packs/GLPI/Integrations/GLPI/GLPI.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (SecInfra) [reply Packs/GLPI/Integrations/GLPI/GLPI.py]: done

File: Packs/GLPI/Integrations/GLPI/GLPI.py [hunk 29]
```diff
@@ -0,0 +1,1008 @@
...
+            else:
+                ticket_fields[arg] = input_arg
+    return ticket_fields
+
+
+def output_format(res, output_type=None, readable=None):
+    if res:
+        if isinstance(res, list):
+            keys = res[0].keys()
+        else:
```
Comment (BEAdi) [inline Packs/GLPI/Integrations/GLPI/GLPI.py:413]:
```suggestion
        elif isinstance(res, str):
            return CommandResults(outputs_prefix='GLPI.' + output_type,
                                  outputs_key_field="id",
                                  outputs=res,
                                  raw_response=res)
        else:
```

**Review by BEAdi — APPROVED** [review]

**Pull request merged**